src/IB/IBFESurfaceMethod.cpp

```diff
@@ -672,19 +1037,274 @@ IBFESurfaceMethod::interpolateVelocity(const int u_data_idx,
+
+                    const Index<NDIM>& ilower = side_boxes[axis].lower();
+                    const Index<NDIM>& iupper = side_boxes[axis].upper();
+
+                    typedef boost::multi_array_types::extent_range range;
+                    boost::const_multi_array_ref<double, NDIM> u_sc_data_array(
+                        u_sc_data->getPointer(axis),
+                        (boost::extents[range(ilower[0] - u_gcw[0], iupper[0] + u_gcw[0] + 1)]
+                                       [range(ilower[1] - u_gcw[1], iupper[1] + u_gcw[1] + 1)]
+#if (NDIM == 3)
+                                       [range(ilower[2] - u_gcw[2], iupper[2] + u_gcw[2] + 1)]
+#endif
+                         ),
+                        boost::fortran_storage_order());
+
+                    for (unsigned int k = 0; k < local_indices.size(); ++k)
+                    {
+                        const int s = local_indices[k];
+                        IBTK::Point x, x_cell, xo, x_cell_o;
+                        const double* const dx = patch_dx;
+                        for (unsigned int d = 0; d < NDIM; ++d)
+                        {
+                            x[d] = x_qp[s * NDIM + d];
+                            ic_center[d] = ilower[d] + boost::math::iround((x[d] - x_lower_axis[d]) / dx[d] - 0.5);
+                            x_cell[d] = x_lower_axis[d] + ((ic_center[d] - ilower[d]) + 0.5) * dx[d];
+                            if (x[d] <= x_cell[d])
+                            {
+                                ic_lower[d] = ic_center[d] - 1;
+                                ic_upper[d] = ic_center[d];
+                            }
+                            else
+                            {
+                                ic_lower[d] = ic_center[d];
+                                ic_upper[d] = ic_center[d] + 1;
+                            }
+
+                            if (x[d] <= x_cell[d])
+                            {
+                                w[d][0] = (x_cell[d] - x[d]) / dx[d];
+                            }
+                            else
+                            {
+                                w[d][0] = 1.0 + (x_cell[d] - x[d]) / dx[d];
+                            }
+                            w[d][1] = 1.0 - w[d][0];
+                            wr[d][0] = +w[d][0];
+                            wr[d][1] = -w[d][1];
+                        }
+                        boost::multi_array<double, NDIM + 1> Ujump(
+                            boost::extents[range(ic_lower[0], ic_upper[0] + 1)][range(ic_lower[1], ic_upper[1] + 1)]
+#if (NDIM == 3)
+                                          [range(ic_lower[2], ic_upper[2] + 1)]
+#endif
+                                          [range(0, NDIM)]);
+
+#if (NDIM == 2)
+                        for (int d = 0; d < NDIM; ++d)
+                        {
+                            LL[d] = (n_qp[s * NDIM] * wr[0][1] + n_qp[s * NDIM + 1] * wr[1][1]) * n_qp[s * NDIM + d];
```

Comment:
If you used a `multi_array` here instead of an `array`, I think that you could replace all of these nearly duplicated statements with loops over sides (L/U) along with the modulo operator.

Reply:
I see what you mean. It can be a lot more compact that way...I'll take a look.